rfcs/0004-transaction.md

```diff
@@ -0,0 +1,291 @@
+
+Another detail worth to be mentioned is that we do not have to track the full keys in `String` on `read_keys` and `write_keys`, which might be not memory efficient. Instead, we could just store the integer hash `KeyFingerPrint` of each key. There might introduce some unnecessary conflicts when different keys happens to got the same hash value, but the probability is extreme low.
+
+Besides the started sequence number of the transaction, we also need to track the sequence number of the transaction committed. During the execution of a transaction, some other transactions might commit, and the current transaction should check the conflicts with these committed transactions.
+
+The conflict check simply checks whether the Read & Write keys of the current transaction are overlapped with the Write keys from related recent committed transactions.
+
+Given an execution history of transactions like:
+
+```
+txn1: [seq:97] <--- ---- ---> [seq:99]
+txn2:     [seq:98]  <--- ---- ---- ---- ---- ---- ---> [seq:105]
+txn3:                         [seq:102] <--- ---> [seq:104]
+txn4:                    [seq:101] <--- ---- ---- ---- ---> [seq:106]
+```
+
+Let's say the current transaction is `txn4`, it should check conflicts with `txn2` and `txn3` during the commit, because `txn2` and `txn3` are committed after `txn4` started. During the conflict checking, `txn4`'s Read & Write keys should not conflict with the Write keys of `txn2` and `txn3`.
+
+It does not need to check conflicts with `txn1`, because `txn1` is committed before `txn4` started. This could tell us when to GC the entry in the `recent_committed_txns` deque: when ALL the running transactions' sequence number are bigger than the committed sequence number of the entry, then the entry could be GCed.
+
+In my opinion, the approach using a global in memory `Oracle` which tracks the recent committed transactions is kinda better than the RocksDB's approach, because it could provide a more accurate conflict check, without coupling with the MemTable size & unpredicatble MemTable flushes. It might introduce some additional memory usage on tracking the recent transactions, but the GC mechamism is also considered as very simple and efficient.
+
+The potential risk of this SSI approach is that ALL the read keys in the running transactions should be tracked in the `read_keys` of the `TransactionState` before commit. If I scanned 1 million rows in a transaction and finally only updated 1 row, all the keys of the 1 million rows scanned should be tracked in side it. This issue could be mitigated by the `KeyFingerPrint` approach, let's say each finger print is 32 bytes, then the 1 million keys would cost roughly 32MB memory, which might be acceptable in most cases.
+
+## Roadmap
+
+1. Implement the WriteBatch API (related with https://github.com/slatedb/slatedb/pull/264).
+2. Add sequence to each key in the storage, and implement the Snapshot API.
```

Comment:
I think this one would need its own RFC. Looking at the previous section, I see
1. In *Snapshot Isolation* we only need to check for overlapping write keys. So, we only to ensure that reads are served from older value (or) a write in this transaction.
2. In "Serializable Snapshot Isolation", in addition, we need to check for  keys read in the transaction as well. 

I am still unclear if we would implement both the isolation levels, or just SSI. If we only implement SSI, is tracking multiple versions for the same key necessary? If a key was updated after we read it, the transaction would abort.

Reply:
Would be good to explicitly specific writing RFC for snapshot API and then implementing is the second step.

Reply:
make sense, i tend to support both, and offer an option to allow user setting an isolation level at transaction or db level.